Impression - Message d'erreurs en format image et résolution de 300dpi

INSPQ:
> Impression
> Message erreur sur les couches impression image 300dpi, mais l’impression fonctionne.

Lorsqu'on a des couches WMS sur la carte et qu'on sélectionne pour l'impression, le format image avec une résolution de 300 dpi, l'application affiche des message d'erreurs "La couche est indisponible pour le moment ou vous n'avez pas les permissions". Par contre, l'impression va tout de même être en succès.

Erreurs dans la console:

> Image.js:194 Error: Image load error
>     at Image.handleError (Image.js:280:14)
>   | setMapImageResolution | @ | print.service.ts:1263
>   | downloadMapImage | @ | print.service.ts:1240
>   | handleFormSubmit | @ | print.component.ts:67

--- a/packages/geo/src/lib/print/print/print.component.html
+++ b/packages/geo/src/lib/print/print/print.component.html
@@ -1,10 +1,10 @@
 <igo-print-form
-  [outputFormat]="outputFormat"
-  [paperFormat]="paperFormat"
-  [orientation]="orientation"
-  [imageFormat]="imageFormat"
-  [resolution]="resolution"
-  [legendPosition]="legendPosition"
+  [outputFormat]="outputFormat()"
+  [paperFormat]="paperFormat()"
+  [orientation]="orientation()"
+  [imageFormat]="imageFormat()"
+  [resolution]="resolution()"
+  [legendPosition]="legendPosition()"
   [disabled$]="disabled$"
   (submit)="handleFormSubmit($event)"
 />

--- a/packages/geo/src/lib/print/print/print.component.ts
+++ b/packages/geo/src/lib/print/print/print.component.ts
@@ -1,4 +1,4 @@
-import { Component, Input, inject } from '@angular/core';
+import { Component, inject, input } from '@angular/core';
 
 import { SecureImagePipe } from '@igo2/common/image';
 
@@ -29,75 +29,20 @@ export class PrintComponent {
 
   public disabled$ = new BehaviorSubject(false);
 
-  @Input()
-  get map(): IgoMap {
-    return this._map;
-  }
-  set map(value: IgoMap) {
-    this._map = value;
-  }
-  private _map: IgoMap;
-
-  @Input()
-  get outputFormat(): PrintOutputFormat {
-    return this._outputFormat;
-  }
-  set outputFormat(value: PrintOutputFormat) {
-    this._outputFormat = value;
-  }
-  private _outputFormat: PrintOutputFormat;
-
-  @Input()
-  get paperFormat(): PrintPaperFormat {
-    return this._paperFormat;
-  }
-  set paperFormat(value: PrintPaperFormat) {
-    this._paperFormat = value;
-  }
-  private _paperFormat: PrintPaperFormat;
-
-  @Input()
-  get orientation(): PrintOrientation {
-    return this._orientation;
-  }
-  set orientation(value: PrintOrientation) {
-    this._orientation = value;
-  }
-  private _orientation: PrintOrientation;
-
-  @Input()
-  get imageFormat(): PrintSaveImageFormat {
-    return this._imageFormat;
-  }
-  set imageFormat(value: PrintSaveImageFormat) {
-    this._imageFormat = value;
-  }
-  private _imageFormat: PrintSaveImageFormat;
-
-  @Input()
-  get legendPosition(): PrintLegendPosition {
-    return this._legendPosition;
-  }
-  set legendPosition(value: PrintLegendPosition) {
-    this._legendPosition = value;
-  }
-  private _legendPosition: PrintLegendPosition;
-
-  @Input()
-  get resolution(): PrintResolution {
-    return this._resolution;
-  }
-  set resolution(value: PrintResolution) {
-    this._resolution = value;
-  }
-  private _resolution: PrintResolution;
+  readonly map = input<IgoMap>(undefined);
+  readonly outputFormat = input<PrintOutputFormat>(undefined);
+  readonly paperFormat = input<PrintPaperFormat>(undefined);
+  readonly orientation = input<PrintOrientation>(undefined);
+  readonly imageFormat = input<PrintSaveImageFormat>(undefined);
+  readonly legendPosition = input<PrintLegendPosition>(undefined);
+  readonly resolution = input<PrintResolution>(undefined);
 
   handleFormSubmit(data: PrintOptions) {
     this.disabled$.next(true);
 
     if (data.isPrintService === true) {
       this.printService
-        .print(this.map, data)
+        .print(this.map(), data)
         .pipe(take(1))
         .subscribe(() => {
           this.disabled$.next(false);
@@ -120,7 +65,7 @@ export class PrintComponent {
 
       this.printService
         .downloadMapImage(
-          this.map,
+          this.map(),
           data.resolution,
           data.imageFormat,
           data.showProjection,

--- a/packages/geo/src/lib/print/shared/print.service.ts
+++ b/packages/geo/src/lib/print/shared/print.service.ts
@@ -12,8 +12,8 @@ import { saveAs } from 'file-saver';
 import html2canvas from 'html2canvas';
 import jsPDF from 'jspdf';
 import { default as JSZip } from 'jszip';
-import { Observable, Subject, forkJoin } from 'rxjs';
-import { map as rxMap } from 'rxjs/operators';
+import { Observable, Subject, forkJoin, fromEvent } from 'rxjs';
+import { catchError, first, map as rxMap, timeout } from 'rxjs/operators';
 
 import { getLayersLegends } from '../../layer/utils/outputLegend';
 import { IgoMap } from '../../map/shared/map';
@@ -25,6 +25,8 @@ import {
   PrintPaperFormat,
   PrintResolution
 } from './print.type';
+
+const PRINT_TIMEOUT = 30000; // 30 secondes
 
 @Injectable({
   providedIn: 'root'
@@ -191,11 +193,11 @@ export class PrintService {
           } else {
             await this.saveDoc(doc);
           }
-        }
 
-        if (status === SubjectStatus.Done || status === SubjectStatus.Error) {
-          this.activityService.unregister(this.activityId);
-          status$.next(SubjectStatus.Done);
+          if (status === SubjectStatus.Done || status === SubjectStatus.Error) {
+            this.activityService.unregister(this.activityId);
+            status$.next(SubjectStatus.Done);
+          }
         }
       }
     );
@@ -718,10 +720,15 @@ export class PrintService {
       resolution,
       viewResolution
     );
-    const status$ = new Subject();
+    const status$ = new Subject<SubjectStatus>();
 
-    let timeout;
-    map.ol.once('rendercomplete', async (event: any) => {
+    this.mapRenderCompleteEvent$(
+      map,
+      mapSize,
+      viewResolution,
+      status$
+    ).subscribe(async (event: any) => {
+      let timeout = undefined;
       const mapCanvas = event.target
         .getViewport()
         .getElementsByTagName('canvas') as HTMLCollectionOf<HTMLCanvasElement>;
@@ -775,7 +782,6 @@ export class PrintService {
         status$.next(status);
       }, 200);
     });
-
     return status$;
   }
 
@@ -1000,7 +1006,7 @@ export class PrintService {
     legendPosition: PrintLegendPosition,
     showNorthArrow: boolean
   ) {
-    const status$ = new Subject();
+    const status$ = new Subject<SubjectStatus>();
     this.activityId = this.activityService.register();
     const translate = this.languageService.translate;
     format = format.toLowerCase();
@@ -1008,7 +1014,12 @@ export class PrintService {
     const initialMapSize = map.ol.getSize() as [number, number];
     const viewResolution = map.ol.getView().getResolution();
 
-    map.ol.once('rendercomplete', async (event: any) => {
+    this.mapRenderCompleteEvent$(
+      map,
+      initialMapSize,
+      viewResolution,
+      status$
+    ).subscribe(async (event: any) => {
       const size = map.ol.getSize();
       const mapCanvas = event.target
         .getViewport()
@@ -1506,5 +1517,29 @@ export class PrintService {
     }
 
     return n * k;
+  }
+
+  private mapRenderCompleteEvent$(
+    map: IgoMap,
+    mapSize: number[],
+    viewResolution: number,
+    status$: Subject<SubjectStatus>
+  ): Observable<unknown> {
+    return fromEvent(map.ol, 'rendercomplete').pipe(
+      timeout(PRINT_TIMEOUT),
+      first(),
+      catchError((error) => {
+        const translate = this.languageService.translate;
+        const message = translate.instant('igo.geo.printForm.timeout', {
+          value: PRINT_TIMEOUT / 1000
+        });
+        error.message += ` - ${message}`;
+        this.messageService.error(message);
+        status$.next(SubjectStatus.Error);
+
+        this.resetOriginalMapSize(map, mapSize, viewResolution);
+        throw error;
+      })
+    );
   }
 }

--- a/packages/geo/src/locale/en.geo.json
+++ b/packages/geo/src/locale/en.geo.json
@@ -381,7 +381,8 @@
           "newpage": "On a new page",
           "newimage": "On a new image"
         },
-        "showNorthArrow": "Show north arrow"
+        "showNorthArrow": "Show north arrow",
+        "timeout": "Print - Some requests reach the maximum time of {{value}} seconds"
       },
       "directions": {
         "buttons": {

--- a/packages/geo/src/locale/fr.geo.json
+++ b/packages/geo/src/locale/fr.geo.json
@@ -380,7 +380,8 @@
           "newpage": "Sur une nouvelle page",
           "newimage": "Sur une nouvelle image"
         },
-        "showNorthArrow": "Afficher la flèche du nord"
+        "showNorthArrow": "Afficher la flèche du nord",
+        "timeout": "Impression - Certaines requêtes atteignent la durée maximale de {{value}} secondes"
       },
       "directions": {
         "buttons": {

